fs/fs_zenfs.cc

```diff
@@ -824,6 +831,79 @@ IOStatus ZenFS::RenameFile(const std::string& source_path,
   return s;
 }
 
+IOStatus ZenFS::LinkFile(const std::string& fname, const std::string& lname,
+                         const IOOptions& options, IODebugContext* dbg) {
+  std::shared_ptr<ZoneFile> src_file(nullptr);
+  IOStatus s;
+
+  Debug(logger_, "LinkFile: %s to %s\n", fname.c_str(), lname.c_str());
+  {
+    std::lock_guard<std::mutex> lock(files_mtx_);
+
+    if (GetFileNoLock(lname) != nullptr)
+      return IOStatus::InvalidArgument("Failed to create link, target exists");
+
+    src_file = GetFileNoLock(fname);
+    if (src_file != nullptr) {
+      src_file->SetLinkName(lname);
+      files_.insert(std::make_pair(lname, src_file));
+      s = SyncFileMetadataNoLock(src_file);
+      if (!s.ok()) {
+        s = src_file->RemoveLinkName(lname);
+        if (!s.ok()) return s;
+        files_.erase(lname);
+      }
+    } else {
+      s = target()->LinkFile(ToAuxPath(fname), ToAuxPath(lname), options, dbg);
+    }
+  }
+  return s;
```

Comment:
We could let go of the lock before going to aux:
```suggestion
  {
    std::lock_guard<std::mutex> lock(files_mtx_);

    if (GetFileNoLock(lname) != nullptr)
      return IOStatus::InvalidArgument("Failed to create link, target exists");

    src_file = GetFileNoLock(fname);
    if (src_file != nullptr) {
      src_file->SetLinkName(lname);
      files_.insert(std::make_pair(lname, src_file));
      s = SyncFileMetadataNoLock(src_file);
      if (!s.ok()) {
        s = src_file->RemoveLinkName(lname);
        if (!s.ok()) return s;
        files_.erase(lname);
      }
      return s;
    } 
  }
  s = target()->LinkFile(ToAuxPath(fname), ToAuxPath(lname), options, dbg);
```

Reply:
Ok we can do that, but  this is still the case with other api's and also it ensures that we do not get into weird race conditions. Do you think holding it affects the performance ?

Reply:
For performance, I don't think it will matter much in the general case. We are not calling this function often. But if something happened to the `target()` that would cause high latency, we would prevent others form accessing `files_` in the time we are waiting for the target filesystem.

I don't think it will be a race. As far as I can see, the only thing that could happen that we check for `fname` in `files_` and it is not there, so we drop the lock. Next thing someone could `OpenWritableFile()` with the same name, and after that we may proceed to call `target()->LinkFile()` even though the source now exist in ZenFS. But that would happen anyway if `OpenWritableFile()` was interleaved after `target()->LinkFile()`.